kernel/objects/signal-listener/objects/SignalListener.cc

```diff
@@ -0,0 +1,250 @@
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ * Copyright 2016 Randolf Rotta, Robert Kuban, and contributors, BTU Cottbus-Senftenberg
+ */
+
+#include "objects/SignalListener.hh"
+#include <new>
+#include "mythos/InvocationBuf.hh"
+#include "util/assert.hh"
+#include "util/PhysPtr.hh"
+#include "objects/ops.hh"
+#include "objects/KernelMemory.hh"
+#include "objects/TypedCap.hh"
+#include "boot/memory-root.hh"
+#include "objects/DebugMessage.hh"
+#include "util/error-trace.hh"
+
+namespace mythos {
+
+  optional<void const*> SignalListener::vcast(TypeId id) const
+  {
+    if (id == typeId<SignalListener>()) return /*static_cast<ExampleObj const*>*/(this);
+    THROW(Error::TYPE_MISMATCH);
+  }
+
+  optional<void> SignalListener::deleteCap(CapEntry&, Cap self, IDeleter& del)
+  {
+    if (self.isOriginal()) {
+      del.deleteObject(del_handle);
+    }
+    RETURN(Error::SUCCESS);
+  }
+
+  void SignalListener::deleteObject(Tasklet* t, IResult<void>* r)
+  {
+    monitor.doDelete(t, [=](Tasklet* t){
+      _mem->free(t, r, this, sizeof(SignalListener));
+    });
+  }
+
+  void SignalListener::invoke(Tasklet* t, Cap self, IInvocation* msg)
+  {
+    monitor.request(t, [=](Tasklet* t){
+        Error err = Error::NOT_IMPLEMENTED;
+        switch (msg->getProtocol()) {
+        case protocol::KernelObject::proto:
+          err = protocol::KernelObject::dispatchRequest(this, msg->getMethod(), self, msg);
+          break;
+        case protocol::SignalListener::proto:
+          err = protocol::SignalListener::dispatchRequest(this, msg->getMethod(), t, self, msg);
+          break;
+        }
+        if (err != Error::INHIBIT) {
+          msg->replyResponse(err);
+          monitor.requestDone();
+        }
+      } );
+  }
+
+  Error SignalListener::invokeBind(Tasklet*, Cap, IInvocation* msg)
+  {
+    auto data = msg->getMessage()->read<protocol::SignalListener::Bind>();
+
+
+    if (data.signalSource() == delete_cap) {
+      _mutex << [this] { _mask = 0; };
+      _source.reset();
+    } else if (data.signalSource() != null_cap) {
+      _mutex << [this] { _mask = 0; };
+      auto err = setSource(msg->lookupEntry(data.signalSource()));
+      if (!err) return err.state();
+    }
+    _mutex << [this, &data] {
+      _mask = data.mask;
+      _context = data.context;
+      _autoResetMask = data.autoResetMask;
+    };
+
+    if (data.keventSink() == delete_cap) { _sink.reset(); }
+    else if (data.keventSink() != null_cap) {
+      auto err = setSink(msg->lookupEntry(data.keventSink()));
+      if (!err) return err.state();
+
+    }
+    _mutex << [this, &data] {
```

Comment:
dupplicate with invokeReset() and maybe also deliver(). Can it be merged into one helper function?